### PR TITLE
Fix pytest hangs in cache cleanup and background shell jobs

### PR DESCRIPTION
**Pull request by basnijholt**
Fix two hangs found while investigating main's pytest runs:

- Prompt-cache allocation can trigger cyclic garbage collection while holding the cache lock, then deadlock when the weakref eviction callback acquires that same lock. Use a reentrant lock and add a subprocess regression that forces this ordering.
- Background shell monitoring can wait for inherited pipe EOF after the parent has exited, preventing bounded reader cleanup. Share the existing foreground exit polling with background monitoring so cleanup can run; existing local and supervisor regressions cover this case.

Validation:
- Full suite: `uv run pytest -m "not requires_matrix" -n 4 --no-cov` — 22,414 passed, 10 skipped.
- The regression fails with the original lock and passes with the fix.
- All 53 prompt-token and schema-cache tests pass.
- All 151 shell tests pass; both inherited-pipe regressions failed before the shell fix.
- Browser and proxy tests pass with sandboxing enabled.
- All repository pre-commit hooks pass.

Current main also passed its rerun: 22,410 tests passed in [pytest run 35528483036](https://github.com/mindroom-ai/mindroom/actions/runs/35528483036).

## Summary by Sourcery

Prevent cache cleanup and background shell monitoring from hanging during process and garbage-collection edge cases.

Bug Fixes:
- Prevent prompt-tool cache deadlocks during garbage-collection eviction callbacks.
- Ensure background shell process cleanup does not hang on inherited pipe EOF.

Tests:
- Add a subprocess regression test covering cache collection during insertion.

## Summary by CodeRabbit

- **Bug Fixes**
  - Prevented occasional hangs when updating cached tool information during memory cleanup.
  - Improved command timeout handling with more responsive deadline enforcement.
  - Fixed background process monitoring so completion is detected reliably, even when output streams remain open.
  - Ensured pending process waits are cancelled cleanly when monitoring ends.
- **Tests**
  - Added coverage for cache updates during garbage collection and reliable process completion behavior.

### CODE REVIEW

File: src/mindroom/shell_execution.py [hunk 1]
```diff
@@ -557,7 +560,7 @@ async def _monitor_process(
 ) -> None:
     """Wait for a backgrounded process to exit and update its record."""
     try:
-        await process.wait()
+        await _await_process_exit(process)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/shell_execution.py:563]: The existing regression already exercises the changed path: `tests/test_shell_output_files.py::test_inherited_pipe_cannot_publish_partial_capture[local-0]` and `[supervisor-0]` both failed with a 10-second timeout before this change, including when rerun in isolation. Both pass after `_monitor_process` uses the shared process-exit wait. The test backgrounds the parent, leaves a child holding the pipe, polls the real handle under a bounded timeout, and asserts an error receipt preserves the existing destination. The focused shell suite passes all 151 tests, and the full suite passes 22,414 tests. No additional duplicate test is needed to cover this regression.
